contracts/adapters/swap/shade-protocol/src/contract.rs

```diff
@@ -0,0 +1,435 @@
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> ContractResult<Response> {
+    // Set contract version
+    // set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Validate entry point contract
+    let checked_entry_point_contract = ContractInfo {
+        address: deps
+            .api
+            .addr_validate(&msg.entry_point_contract.address.to_string())?,
+        code_hash: msg.entry_point_contract.code_hash,
+    };
+
+    ENTRY_POINT_CONTRACT.save(deps.storage, &checked_entry_point_contract)?;
+
+    let checked_shade_router_contract = ContractInfo {
+        address: deps
+            .api
+            .addr_validate(&msg.shade_router_contract.address.to_string())?,
+        code_hash: msg.shade_router_contract.code_hash,
+    };
+
+    SHADE_ROUTER_CONTRACT.save(deps.storage, &checked_shade_router_contract)?;
+
+    VIEWING_KEY.save(deps.storage, &msg.viewing_key)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute(
+            "entry_point_contract_address",
+            checked_entry_point_contract.address.to_string(),
+        )
+        .add_attribute(
+            "shade_router_contract_address",
+            msg.shade_router_contract.address,
+        ))
+}
+
+///////////////
+/// RECEIVE ///
+///////////////
+
+// Receive is the main entry point for the contract to
+// receive cw20 tokens and execute the swap
+pub fn receive_snip20(
+    deps: DepsMut,
+    env: Env,
+    mut info: MessageInfo,
+    snip20_msg: Snip20ReceiveMsg,
+) -> ContractResult<Response> {
+    // Set the sender to the originating address that triggered the cw20 send call
+    // This is later validated / enforced to be the entry point contract address
+    info.sender = deps.api.addr_validate(&snip20_msg.sender.to_string())?;
+
+    match snip20_msg.msg {
+        Some(msg) => match from_binary(&msg)? {
+            Snip20HookMsg::Swap { operations } => {
+                execute_swap(deps, env, info, operations, snip20_msg.amount)
+            }
+        },
+        None => Ok(Response::default()),
+    }
+}
+
+///////////////
+/// EXECUTE ///
+///////////////
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> ContractResult<Response> {
+    match msg {
+        ExecuteMsg::Receive(snip20_msg) => receive_snip20(deps, env, info, snip20_msg),
+        ExecuteMsg::TransferFundsBack {
+            swapper,
+            return_denom,
+        } => Ok(execute_transfer_funds_back(
+            deps,
+            env,
+            info,
+            swapper,
+            return_denom,
+        )?),
+        // Tokens must be registered before they can be swapped
+        ExecuteMsg::RegisterTokens { contracts } => register_tokens(deps, env, contracts),
```

Comment:
It's intended for this method to be callable by anyone, yeah? In practice will Shade register all the tokens once the contracts are deployed?

Reply:
We certainly can, it will be good for you guys to be able to do it as well but I'd be happy to seed the data  with all the tokens we know about